progs/jubjub.py

```diff
@@ -0,0 +1,315 @@
+        self.curve = curve
+        self.xs = xs
+        self.ys = ys
+
+    async def __init(self):
+        if not await(self.curve.contains_shared_point(self.context, self)):
+            raise Exception(
+                f"Could not initialize Point {self}-- \
+                does not sit on given curve {self.curve}")
+
+    @staticmethod
+    async def create(context: Mpc, xs, ys, curve=Jubjub()):
+        point = SharedPoint(context, xs, ys, curve)
+        await(point.__init())
+        return point
+
+    @staticmethod
+    async def from_point(context: Mpc, p: Point) -> 'SharedPoint':
+        if not isinstance(p, Point):
+            raise Exception(f"Could not create shared point-- p ({p}) is not a Point!")
+
+        return await(SharedPoint.create(context, context.Share(p.x), context.Share(p.y)))
+
+    def __str__(self) -> str:
+        return f"({self.xs}, {self.ys})"
+
+    def __repr__(self) -> str:
+        return str(self)
+
+    async def neg(self):
+        return await(SharedPoint.create(self.context,
+                                        await(self.context.Share(-1) * self.xs),
+                                        self.ys,
+                                        self.curve))
+
+    async def add(self, other: 'SharedPoint') -> 'SharedPoint':
+        if self.curve != other.curve:
+            raise Exception("Can't add points on different curves!")
+        elif self.context != other.context:
+            raise Exception("Can't add points from different contexts!")
+        elif isinstance(other, Ideal):
+            return self
+        elif not isinstance(other, SharedPoint):
+            raise Exception(
+                f"Could not add other point-- not an instance of SharedPoint")
+
+        x1, y1, x2, y2 = self.xs, self.ys, other.xs, other.ys
+
+        one = self.context.Share(1)
+        y_prod = await(y1 * y2)
+        x_prod = await(x1 * x2)
+
+        # d_prod = d*x1*x2*y1*y2
+        d_prod = await(await(self.context.Share(self.curve.d) * x_prod) * y_prod)
+
+        # x3 = ((x1*y2) + (y1*x2)) / (1 + d*x1*x2*y1*y2)
+        x3 = await((await(x1 * y2) + await(y1 * x2)) / (one + d_prod))
+
+        # y3 = ((y1*y2) + (x1*x2)) / (1 - d*x1*x2*y1*y2)
+        y3 = await((y_prod + x_prod) / (one - d_prod))
+
+        return await(SharedPoint.create(self.context, x3, y3, self.curve))
+
+    async def sub(self, other: 'SharedPoint') -> 'SharedPoint':
+        return self.add(await(other.mul(-1)))
+
+    async def mul(self, n: int) -> 'SharedPoint':
+        if not isinstance(n, int):
+            raise Exception("Can't scale a SharedPoint by something which isn't an int!")
+
+        if n < 0:
+            return await(await(self.mul(-1)).mul(-n))
+        elif n == 0:
+            return Ideal(self.curve)
+
+        Q = self
+        R = self if n & 1 == 1 else Ideal(self.curve)
+
+        i = 2
+        while i <= n:
+            Q = await(Q.double())
+
+            if n & i == i:
```

Comment:
I think this logic can be simplified. It's currently split between `R = self if n & 1 == 1 else Ideal(self.curve)` which only considers the least significant bit of `n` and the logic in this loop which handles the rest.

Reply:
You're probably right-- I'll try to do this: this is just directly copied from the existing jubjub code.